Substation maintenance

Partmedia opened the pull request:
## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->

Add a substation maintenance mechanic. Substations slowly burn oil while they're running. The more power the substation is delivering, the more oil it burns. If a substation is not maintained by periodic oil refilling, it will take damage until it fails and needs to be rebuilt.

This mechanic is designed to give mid-round engineers some work to do. Instead of setting up everything at round start and lounging for the rest of the shift, responsible engineers will need to visit each substation periodically and check up on their oil levels. This is also better than a random event, because this mechanic allows good engineers to preemptively maintain their equipment and prevent failures.

In support of this change:

- Add a "can of oil" prototype and give engineers some at at round start in the electrical locker.
- Disable the BreakerFlip random event.
- While here, adjust the mapping suffix of the substation prototype with correct units (MW to MJ).

A nice side effect of this change: responsible engineers cannot "just" fix power by bridging MV or LV segments, because that would overload a substation and cause it to burn oil at a faster rate. If the substation is not subsequently serviced, it will fail much faster because it is overloaded.

## Screenshots
<!-- If applicable, add screenshots to showcase your PR. If your PR is a visual change, add
screenshots or it's liable to be closed by maintainers. -->
![submaint](https://user-images.githubusercontent.com/3229565/193165035-1cf0d45b-2188-44af-8ba2-4218a092f9ba.png)

## FAQ
**Why not just have mice bite random wires or have random wires burn out?**
Substations are easier to locate and cannot be under walls.

**Why oil in substations? That seems fake.**
It's a [real thing](https://electrical-engineering-portal.com/oil-filled-equipment-in-substation).

**How long will it take for a substation to fail?**
At round start, a typical heavily-loaded substation has a load of around 16 kW. The oil consumption rate is: exp(16kW/5kW)/60 = 0.4u per minute. Since substations start filled with 30u, it will finish consuming its oil, if not maintained, in 30u/0.4u = 73 minutes.

Then, substations have a 10% chance of taking 10 damage (out of 100) every 10 seconds. That means on average, it will take 10 damage every 1/0.1 * 10 seconds = 100 seconds. Since it takes 10 hits to destroy the substation, it takes 1000 seconds = 17 minutes on average.

So, the time to expected failure is 73 + 17 = **90 minutes**. Of course, more heavily loaded substations will fail sooner, and lightly loaded substations will not fail in a reasonable amount of time. These numbers, of course, can be tuned.

**Won't all the substations fail at the same time and everything will be terrible?**
No. Substations consume oil at a rate based on the amount of power they are supplying. The first substation to fail will likely be the substation that is loaded the most. There is a good distribution of substation loads across all the stations available:

Bagel:
![bagel](https://user-images.githubusercontent.com/3229565/193165260-d9c0f4b1-15eb-44e2-938b-a7c65b478fc3.png)

Maraton:
![marathon](https://user-images.githubusercontent.com/3229565/193165267-4fa7e763-4f70-4b8a-b3bf-b66a0ee839fb.png)

Meta:
![meta](https://user-images.githubusercontent.com/3229565/193165270-b18b5b8e-22e9-4754-a675-322554a2e1a0.png)

Omega:
![omega](https://user-images.githubusercontent.com/3229565/193165273-dd4450ee-9f65-4a9a-bbd8-1c29c78f3879.png)

Packed:
![packed](https://user-images.githubusercontent.com/3229565/193165276-9e7fd674-d636-4a07-8e79-5918cccede2a.png)

Pillar:
![pillar](https://user-images.githubusercontent.com/3229565/193165278-f812a825-91ec-40a8-b6a6-44d818a122ef.png)

Saltern:
![saltern](https://user-images.githubusercontent.com/3229565/193165283-2a0f289c-08b9-43c0-9369-431219ad77d1.png)

**What happens if someone puts water in the substation?**
That would be traitorous activity. The examine text would show that the substation is still well-filled, but once all the oil is used up, the substation will start taking damage (despite it appearing to be well-filled). It does not (currently) accelerate damage.

**Changelog**
:cl: notafet
- add: Substations now require maintenance to prevent eventual failure.

metalgearsloth commented on Content.Server/Power/EntitySystems/MaintenanceOilSystem.cs:
```diff
@@ -0,0 +1,118 @@
+    [Dependency] private readonly IRobustRandom _robustRandom = default!;
+    [Dependency] private readonly SolutionContainerSystem _solutionContainerSystem = default!;
+
+    private float _updateTimer = 0.0f;
+    private const float UpdateTime = 10.0f; // 10 seconds. Can't be much lower or FixedPoint2 isn't precise enough.
```
Don't store stateful data on systems.

Partmedia replied:
This state is used for the system to keep track of when to periodically update. The same pattern is used by ToolSystem.Welder, BarotraumaSystem, FlammableSystem, AdminLogManager, CrewMonitoringConsoleSystem, PowerMonitoringConsoleSystem, SmokingSystem, AdvertiseSystem, and ResearchSystem. Can you suggest an alternative?

theashtronaut replied:
store it on the component, in the update loop you can subtract / add time and compare on a component basis. This would also allow update staggering in the future if desired. take a look at BloodstreamSystem or GasAnalyzerSystem (I based it on BloodstreamSystem) as those both do this

Partmedia replied:
It's a trivial change, and if you insist I'll go ahead and make it. Staggering is not necessary for this system since it is probabilistic on a substation-to-substation basis, anyway. Looping through all substations and and individually checking and bumping their update time seems unnecessary and wasteful, especially since this only needs to update once every 10 seconds.

metalgearsloth reviewed (changes requested):
You need comments because "MaintenanceOil" is not self-descriptive.

Also:
What happens if someone puts water in the substation?
That would be traitorous activity. The examine text would show that the substation is still well-filled, but once all the oil is used up, the substation will start taking damage (despite it appearing to be well-filled). It does not (currently) accelerate damage.

Allowing this for griefing seems like a bad idea considering ingame you have no way of knowing the actual reagents inside?

theashtronaut commented:
> What happens if someone puts water in the substation?
That would be traitorous activity. The examine text would show that the substation is still well-filled, but once all the oil is used up, the substation will start taking damage (despite it appearing to be well-filled). It does not (currently) accelerate damage.

Is there a way to detect this at all? If the transformer was fully filled with water is there a way to 'drain' the transformer of the water and replace it with oil again?

Partmedia commented:
> > What happens if someone puts water in the substation?
> > That would be traitorous activity. The examine text would show that the substation is still well-filled, but once all the oil is used up, the substation will start taking damage (despite it appearing to be well-filled). It does not (currently) accelerate damage.
> 
> Is there a way to detect this at all? If the transformer was fully filled with water is there a way to 'drain' the transformer of the water and replace it with oil again?

Currently, deconstruct and reconstruct the substation. If someone poured water into your car's gas tank, how'd you get it out? We can consider adding a siphon, but see below.

I could perhaps add examine text for when the substation is taking damage, so that you could tell it's out of oil despite the fill line indicating otherwise.

I didn't consider this issue important enough, because if someone really wanted to sabotage your substation and has already gotten into the substation vault, they could just blow it up.

metalgearsloth commented:
If you're meleeing a substation it's going to blow up in your face vs some indeterminate time later.

metalgearsloth commented:
At a minimum make it require access so any greytide walking past can't mess with it, but in future the examine should disclose what's in there as someone who isn't familiar with the future isn't going to know why a substation spontaneously broke.

Partmedia commented:
> If you're meleeing a substation it's going to blow up in your face vs some indeterminate time later.

How about blowing the substation up immediately after a sufficient quantity of non-oil reagent has been added? This makes sense because it would be "shorting out" the substation.

> At a minimum make it require access so any greytide walking past can't mess with it, but in future the examine should disclose what's in there as someone who isn't familiar with the future isn't going to know why a substation spontaneously broke.

So, most transformers are behind engineering-secured doors. There are a handful of transformers out in the open on Packed, and maybe other stations. Isn't this really a mapping issue? Just like how we don't put trigger locks on guns, we just put them behind an armory?

GoodWheatley commented:
Just have a question about this, is there any way for engineering to obtain more oil in case they run out? If there's no way to get more a traitor engineer could just steal all the oil cans and start filling substations with water, and there's nothing the engineering team could do to fix it.

Partmedia commented:
> Just have a question about this, is there any way for engineering to obtain more oil in case they run out? If there's no way to get more a traitor engineer could just steal all the oil cans and start filling substations with water, and there's nothing the engineering team could do to fix it.

Chemists can make oil, so if the stock is depleted, the chemists can make more.

Of course, traitor engineers could just blow up substations with a brick of C4. But it sounds like multiple people are concerned about not being able to spot damage, so I will think of a solution. Either changing the examine text, make some other sign of failure visible, or just make it instant blow up.

lapatison commented:
> > Just have a question about this, is there any way for engineering to obtain more oil in case they run out? If there's no way to get more a traitor engineer could just steal all the oil cans and start filling substations with water, and there's nothing the engineering team could do to fix it.
> 
> Chemists can make oil, so if the stock is depleted, the chemists can make more.
> 
> Of course, traitor engineers could just blow up substations with a brick of C4. But it sounds like multiple people are concerned about not being able to spot damage, so I will think of a solution. Either changing the examine text, make some other sign of failure visible, or just make it instant blow up.

![image](https://user-images.githubusercontent.com/100279397/193364768-f4365115-e054-482f-bd05-2028159c6db3.png)
You mean this oil?

theashtronaut commented:
> Of course, traitor engineers could just blow up substations with a brick of C4. But it sounds like multiple people are concerned about not being able to spot damage, so I will think of a solution. Either changing the examine text, make some other sign of failure visible, or just make it instant blow up.

I do like the concept of subtle, delayed sabotage. I just also want it to be something that could be noticed and stopped without too much annoyance as well.

Maybe make it so you can drain the substation's contents, but if you drain too much oil while it is active then it would immediately catch fire / harm the person working on it. This way the antag can drain some and replace with water for the delayed sabotage, drain all and replace with water for immediate but with consequences on themselves, and the engineer can fully repair / replace the oil by powering the substation down (by cutting the HV wire first) to remove any contaminants. That also means during the complete replace you lose power to whatever it's connected to, just like with the rebuild.

Having a way to tell what's in the substation so you can tell when there is non-oil would be handy, since otherwise it would just appear like it's full but not working which could be frustrating to figure out the actual problem is there isn't actually oil in it.

Partmedia commented:
> > > Just have a question about this, is there any way for engineering to obtain more oil in case they run out? If there's no way to get more a traitor engineer could just steal all the oil cans and start filling substations with water, and there's nothing the engineering team could do to fix it.
> > 
> > 
> > Chemists can make oil, so if the stock is depleted, the chemists can make more.
> > Of course, traitor engineers could just blow up substations with a brick of C4. But it sounds like multiple people are concerned about not being able to spot damage, so I will think of a solution. Either changing the examine text, make some other sign of failure visible, or just make it instant blow up.
> 
> ![image](https://user-images.githubusercontent.com/100279397/193364768-f4365115-e054-482f-bd05-2028159c6db3.png) You mean this oil?

I mean this oil. I can fix the description. Oil is not actually used in any recipe, and it doesn't make sense that chefs are cooking with something that contains welding fuel.

Partmedia commented:
> > Of course, traitor engineers could just blow up substations with a brick of C4. But it sounds like multiple people are concerned about not being able to spot damage, so I will think of a solution. Either changing the examine text, make some other sign of failure visible, or just make it instant blow up.
> 
> I do like the concept of subtle, delayed sabotage. I just also want it to be something that could be noticed and stopped without too much annoyance as well.
> 
> Maybe make it so you can drain the substation's contents, but if you drain too much oil while it is active then it would immediately catch fire / harm the person working on it. This way the antag can drain some and replace with water for the delayed sabotage, drain all and replace with water for immediate but with consequences on themselves, and the engineer can fully repair / replace the oil by powering the substation down (by cutting the HV wire first) to remove any contaminants. That also means during the complete replace you lose power to whatever it's connected to, just like with the rebuild.
> 
> Having a way to tell what's in the substation so you can tell when there is non-oil would be handy, since otherwise it would just appear like it's full but not working which could be frustrating to figure out the actual problem is there isn't actually oil in it.

I like this suggestion.

Any suggestions for how to tip the substation and drain things from it? DrainableComponent doesn't quite to do what we want it to do, since Substation doesn't have ItemComponent.

theashtronaut commented:
> 
> I like this suggestion.
> 
> Any suggestions for how to tip the substation and drain things from it? DrainableComponent doesn't quite to do what we want it to do, since Substation doesn't have ItemComponent.

My guess would be to look into how other tanks work, like the welding fuel / water tanks. I believe they use ReagentTank. Only problem is I don't think those tanks let you fill them. If that's a constraint on ReagentTank then maybe there is a better way to keep track of a solution? I haven't really looked into it too much yet.

Partmedia commented:
- Updated to fix merge conflict (and force pushed)
- Added drainable component so that substation contents can be sampled and inspected or drained. Of course, don't drain the substation fully while it's on. Call your local chemist to figure out what you actually sampled.
- Require maintenance for wallmount substations as well.

mirrorcult commented:
very much like this as a subtle mechanic. does having it both be refillable/drainable even work? i remember there being some issues with that previously

Partmedia commented:
> very much like this as a subtle mechanic. does having it both be refillable/drainable even work? i remember there being some issues with that previously

Looks like it does not. I'll take a look at those systems and see what it'll take for that to work (probably adding verbs instead of using only OnInteract).

Partmedia commented:
Overcome by events.
